handler/mosn/filter.go

```diff
@@ -0,0 +1,179 @@
+		httpwasm.GuestConfig([]byte(conf)),
+		httpwasm.Logger(func(ctx context.Context, s string) {
+			log.Proxy.Infof(ctx, "wasm: %s", s)
+		}))
+	if err != nil {
+		return nil, err
+	}
+
+	return &filterFactory{
+		rt: rt,
+	}, nil
+}
+
+type filterFactory struct {
+	rt *internalhandler.Runtime
+}
+
+func (f filterFactory) CreateFilterChain(context context.Context, callbacks api.StreamFilterChainFactoryCallbacks) {
+	fr := &filter{
+		rt: f.rt,
+		ch: make(chan error, 1),
+	}
+	callbacks.AddStreamReceiverFilter(fr, api.BeforeRoute)
+	callbacks.AddStreamSenderFilter(fr, api.BeforeSend)
+}
+
+type filter struct {
+	rt *internalhandler.Runtime
+
+	features handler.Features
+
+	receiverFilterHandler api.StreamReceiverFilterHandler
+
+	reqHeaders api.HeaderMap
+	reqBody    api.IoBuffer
+
+	respStatus  int
+	respHeaders api.HeaderMap
+	respBody    []byte
+
+	nextCalled bool
+	ch         chan error
+}
+
+func (f *filter) OnReceive(ctx context.Context, headers api.HeaderMap, body api.IoBuffer, _ api.HeaderMap) api.StreamFilterStatus {
+	ctx = context.WithValue(ctx, filterKey{}, f)
+
+	f.reqHeaders = headers
+	f.reqBody = body
+
+	go func() {
+		f.ch <- f.rt.Handle(ctx)
+	}()
+
+	err := <-f.ch
+	if err != nil {
+		log.Proxy.Errorf(ctx, "wasm error: %v", err)
+	}
+
+	if f.nextCalled {
+		return api.StreamFilterContinue
+	}
+
+	// TODO(anuraaga): All mosn filter examples pass the request headers when sending a hijack reply. Trying to send
+	// f.respHeaders causes the hijack to be ignored. Figure out why.
+	if respBody := f.respBody; len(respBody) > 0 {
+		f.receiverFilterHandler.SendHijackReplyWithBody(f.respStatus, headers, string(respBody))
+	} else {
+		f.receiverFilterHandler.SendHijackReply(f.respStatus, headers)
+	}
+	return api.StreamFilterStop
+}
+
+func (f *filter) SetReceiveFilterHandler(handler api.StreamReceiverFilterHandler) {
+	f.receiverFilterHandler = handler
+}
+
+func (f *filter) OnDestroy() {
+}
+
+func (f *filter) Append(ctx context.Context, headers api.HeaderMap, buf api.IoBuffer, trailers api.HeaderMap) api.StreamFilterStatus {
+	if !f.nextCalled {
+		// TODO(anuraaga): All mosn filter examples pass the request headers when sending a hijack reply. We replace
+		// with response headers here until fixing that.
+		// There is no headers.Clear() for some reason.
+		headers.Range(func(key, value string) bool {
+			headers.Del(key)
+			return true
+		})
+		if f.respHeaders != nil {
+			f.respHeaders.Range(func(key, value string) bool {
+				headers.Set(key, value)
+				return true
+			})
+		}
+		return api.StreamFilterStop
+	}
+
+	ctx = context.WithValue(ctx, filterKey{}, f)
+
+	if f.respHeaders != nil {
+		f.respHeaders.Range(func(key, value string) bool {
+			headers.Set(key, value)
+			return true
+		})
+	}
+	f.respHeaders = headers
+	f.respBody = buf.Bytes()
+
+	f.ch <- nil
+	err := <-f.ch
+	if err != nil {
+		log.Proxy.Errorf(ctx, "wasm error: %v", err)
+		return api.StreamFilterContinue
+	}
+
+	// BUG: mosn doesn't use psuedoheaders.
```

Comment:
Realized I have this bug still left, if it'll make other PRs hard will check it out before merge